src/writer/creator.cpp

```diff
@@ -383,13 +393,27 @@ namespace zim
       _write(out_fd, reinterpret_cast<const char*>(digest), 16);
     }
 
+    void Creator::checkError()
+    {
+      if (data->m_errored) {
+        throw std::runtime_error("Creator is in error state");
+      }
+      std::lock_guard<std::mutex> l(data->m_exceptionLock);
+      if (data->m_exceptionSlot) {
+        std::cerr << "ERROR Detected" << std::endl;
+        data->m_errored = true;
+        std::rethrow_exception(data->m_exceptionSlot);
```

Comment:
I wonder if rethrowing the exception intercepted in a worker thread makes sense. Because of the asynchronous nature of such error handling (the error originates in one context but is reported in another) it may lead to confusion. What about adding an `AsyncError` exception type that will link to the intercepted exception via a `std::exception_ptr` data member?

Reply:
It looks like a good idea but it seams that we can do nothing but store and rethrow a `std::exception_ptr`.
So it is useless to store in in a AsyncError, especially we want to know the error message of the original exception and a `AsyncError` we may throw cannot access it.
(Or maybe you know how to)

Reply:
> Edit: My bad, we can. We can rethrow it in a try/catch.

Exactly. That's what I had in mind. And I also think that it will be useful to utilize that trick to extract the error message from the intercepted exception in `AsyncError::what()` (or in the constructor of `AsyncError`) and construct an informative error message. Then only code that needs to know the exact type of the intercepted exception will have to have an extra `try{}catch(){}` inside `catch(const AsyncError& ex){}`. Yet that will make sense only in the following cases:

1. when it's possible to recover from that particular error situation
2. or when there are necessary clean-up actions to be performed specific to such a failure